Ensure nftables unit test parity with iptables

npinaeva opened the pull request:
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
https://github.com/kubernetes/kubernetes/issues/122572
Add missing packet tracing unit tests

#### Special notes for your reviewer:
Unit test rules are coming from https://github.com/kubernetes/kubernetes/blob/b98e47dd80166935117f890b437c8e30a91da173/pkg/proxy/nftables/proxier_test.go#L293, but all chain declarations were put in the beginning to avoid "no such chain" error

To ensure removed parsing functions are not used, we can go to https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/nftables/proxier.go and search for "dport" usages.
You will see that there are 2 possible following symbols:
- `vmap`
- `svcInfo.Port()`
so the removed regex's will never be matched

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

tnqn commented on pkg/proxy/nftables/helpers_test.go:
```diff
@@ -815,3 +815,196 @@ func Test_diffNFTablesChain(t *testing.T) {
 	}
 }
+
+// This tests tracePackets against static data, just to make sure we match things in the
```
There is no func called `tracePackets`. Since it's copied from iptables test, should we correct both places to `tracePacket`?

npinaeva replied:
good point, let me fix this

tnqn commented on pkg/proxy/nftables/helpers_test.go:
```diff
@@ -815,3 +815,196 @@ func Test_diffNFTablesChain(t *testing.T) {
+			output:   "10.180.0.4:80, 10.180.0.5:80",
+		},
+		{
+			name:     "LOCAL, KUBE-MARK-MASQ",
```
`KUBE-MARK-MASQ` is an iptables chain but not a nftables chain, should the name just be `LOCAL, MASQUERADE`?

danwinship replied:
"LOCAL" (and "KUBE-MARK-MASQ") here refer to the specific iptables rule that it's testing in the iptables example ("`... --src-type LOCAL -j KUBE-MARK-MASQ`"). As it happens, this exactly corresponds to the "`... fib saddr type local jump mark-for-masquerade`" in this test, so maybe "local, mark-for-masquerade" as a name.

In general, the goal of the test case is to make sure that each of the regexes in `helpers_test.go` and their corresponding `switch` clauses in `runChain` gets tested. So you should confirm that. You may need some slightly different tests than are in the iptables version since the rules don't all work the same way...

danwinship commented on go.mod:
```diff
@@ -121,7 +121,7 @@ require (
 	k8s.io/system-validators v1.8.0
 	k8s.io/utils v0.0.0-20230726121419-3b25d923346b
-	sigs.k8s.io/knftables v0.0.14
+	sigs.k8s.io/knftables v0.0.15
```
You should do the knftables bump as a separate commit

danwinship reviewed (commented):
> but all chain declarations were put in the beginning to avoid "no such chain" error

actually, now that we have `ParseDump`, can you fix the ordering of the dump in `TestOverallNFTablesRules` and then call `ParseDump` on it as part of the unit test, to make sure that people keep the rules in the right order in the future as things change?

(as a separate commit from the new unit test)

npinaeva committed:
kube-proxy: rename iptables unit test to match tested function name.

Signed-off-by: Nadia Pinaeva <n.m.pinaeva@gmail.com>

npinaeva committed:
kube-proxy: ensure nftables unit test parity with iptables.
Add packet tracing unit tests for ipv4 and ipv6.
Remove unreachable code from runChain, since some of the parsed rules
are never generated by the proxy implementation.

Signed-off-by: Nadia Pinaeva <n.m.pinaeva@gmail.com>

danwinship commented on pkg/proxy/nftables/helpers_test.go:
```diff
@@ -815,3 +776,321 @@ func Test_diffNFTablesChain(t *testing.T) {
+			output:   "10.180.0.4:80, 10.180.0.5:80",
+		},
+		{
+			name:     "LOCAL, mark for masquerade",
```
"local, mark for masquerade"
("LOCAL" in all caps refers to part of the iptables rule, but the corresponding nftables rule uses "local")

danwinship commented on pkg/proxy/nftables/helpers_test.go:
```diff
@@ -815,3 +776,321 @@ func Test_diffNFTablesChain(t *testing.T) {
+			masq:     true,
+		},
+		{
+			name:     "DROP",
```
(It's fine to keep this in all-caps though since the tracePacket output is still "DROP" even though the nftables rule is "drop").

danwinship commented on go.mod:
```diff
@@ -120,7 +120,7 @@ require (
 	k8s.io/system-validators v1.8.0
 	k8s.io/utils v0.0.0-20230726121419-3b25d923346b
-	sigs.k8s.io/knftables v0.0.14
+	sigs.k8s.io/knftables v0.0.16-0.20240228165918-78f68d9b296e
```
So it looks like we're good to tag this release then? (Also, this commit should be before the commit that depends on it, not after.)

danwinship replied:
ok, `v0.0.16` is tagged

npinaeva committed:
Update knftables to v0.0.16

Signed-off-by: Nadia Pinaeva <n.m.pinaeva@gmail.com>

npinaeva commented on staging/src/k8s.io/kms/internal/plugins/_mock/go.work.sum:
`make update` did that for some reason, seems to be gone now

danwinship commented:
/lgtm

/assign @thockin
for vendor update to latest knftables
(which is now pulling in non-code files (README.md, etc) that it didn't used to pull in? Is that expected with the new workspaces-based scripts?)

npinaeva commented:
/retest

thockin commented:
/approve

cici37 commented:
/remove-sig api-machinery
